crates/sparrow-instructions/src/evaluators/cast.rs

```diff
@@ -89,6 +104,34 @@ impl CastEvaluator {
                     input.data_type()
                 )),
             },
+            // Arrow does not support casting directly from duration to duration types, but it does
+            // support casting to the underlying data type (int64) then back to any duration type.
+            // This hacks that cast by doing the multiplication manually then casting back to the
+            // desired type.
+            (DataType::Duration(TimeUnit::Second), to_type @ DataType::Duration(tu)) => match tu {
```

Comment:
These can probably be simplified in a few ways:

1. We can cast from duration to int64array and vice versa for free.
2. This means instead of using `arity::unary` to do the multiplication we can just use the multiply by a scalar kernel.
3. Instead of special casing, we can compute the multiplier -- if we know how many nanoseconds are in each unit, then `from_nanos / to_nanos` gives us the conversion. We could maybe check which is greater and use division if going the other direction.

I think putting that together gives us something like:

```
fn time_unit_nanos(unit: &TimeUnit) -> i64 {
  match unit {
    TimeUnit::Second => 1_000_000_000,
    TimeUnit::Millisecond => 1_000_000,
    TimeUnit::Microsecond => 1_000,
    TimeUnit::Nanosecond => 1
  }
}

fn cast_duration(input: &dyn Array, from: &TimeUnit, to: &TimeUnit) -> ArrayRef {
  let input: ArrayRef = arrow::compute::cast(input, DataType::Int64Array)?;
  let input: &Int64Array = input.as_primitive();
  
  let from = time_unit_nanos(from);
  let to = time_unit_nanos(to);
  let result: ArrayRef = if from < to {
    arrow::compute::kernels::numeric::mul(&input, to / from)?
  } else if from > to {
    arrow::compute::kernels::numeric::div(&input, from / to)?
  };
  let result = arrow::compute::cast(result.as_ref(), DataType::Duration(*to))?;
  Ok(result)
}
```

Then you can put more of your comment on the `cast_duration` function. That will avoid needing as many special cases.